lib/collection/src/collection_manager/segments_updater.rs

```diff
@@ -130,15 +143,43 @@ pub(crate) fn set_payload(
     payload: &Payload,
     points: &[PointIdType],
 ) -> CollectionResult<usize> {
-    let updated_points =
-        segments.apply_points_to_appendable(op_num, points, |id, write_segment| {
-            write_segment.set_payload(op_num, id, payload)
-        })?;
+    let updated_points = segments.apply_points_and_reindex(
+        op_num,
+        points,
+        |id, write_segment| write_segment.set_payload(op_num, id, payload),
+        |segment| safe_to_set_payload(&segment.get_indexed_fields(), payload),
+    )?;
 
     check_unprocessed_points(points, &updated_points)?;
     Ok(updated_points.len())
 }
 
+fn safe_to_set_payload(
+    indexed_fields: &HashMap<PayloadKeyType, PayloadFieldSchema>,
+    payload: &Payload,
+) -> bool {
+    // Set is safe when the provided payload doesn't intersect with any of the indexed fields.
+    // Note that if we have, e.g., an index on "a.c" and the payload being set is `{"a": {"b": 1}}`,
+    // it is not safe because the whole value of "a" is being replaced.
+    indexed_fields
+        .keys()
+        .all(|path| !payload.0.contains_key(get_path_head(path)))
+}
+
+fn safe_to_delete_payload_keys(
+    indexed_fields: &HashMap<PayloadKeyType, PayloadFieldSchema>,
+    keys_to_delete: &[PayloadKeyType],
```

Comment:
Note: accepting the whole `HashMap` is not necessarly (an `Iterator<Item = &PayloadKeyType>` would suffice), but I did it here explicitly so this function arguments won't get mixed up by accident.